### PR TITLE
Fix playlists not allowing entry with partial mods

### LINKED ISSUE #16978: Classic Customization in Playlist loses partial toggles
### Discussed in https://github.com/ppy/osu/discussions/16968

<div type='discussions-op-text'>

<sup>Originally posted by **itotallygetthat** February 23, 2022</sup>
[network.log](https://github.com/ppy/osu/files/8121902/network.log)
[runtime.log](https://github.com/ppy/osu/files/8121903/runtime.log)
[performance.log](https://github.com/ppy/osu/files/8121904/performance.log)
[updater.log](https://github.com/ppy/osu/files/8121905/updater.log)
[database.log](https://github.com/ppy/osu/files/8121906/database.log)

Lazer 2022.217.0

Game is giving an error when attempting to start a map which is in a playlist if this condition is met:
- map was added to the playlist with the Classic options enabled, customized so that only some of the Classic options are toggled on.

The error log mentions "Current Mods do not match PlaylistItem's RequiredMods".

As of a week ago (version 2022.205.0), when some Classic options were toggled, the hover text would say "XYZ option set to False," for each Classic option toggled to off.  Today (version 2022.217.0) the same hover text just says "Classic".

A map with full Classic options starts ok.
A map without Classic enabled starts ok.
Partial Classic not in a playlist starts ok.
See the playlist "partial classic test" for an example.  (https://osu.ppy.sh/multiplayer/rooms/150654)

My guess is that the client is no longer getting the Classic customization options loaded in properly in the playlist map selection menu from the server's playlist, but the server's playlist has the correct Classic customization options stored properly.
And that something since the 205 update led to this behavior.</div>

### PR DESCRIPTION
**Pull request by smoogipoo**
Fixes https://github.com/ppy/osu/issues/16978

There are two issues here.

The first is that APIMod would store the `Bindable` rather than the value, which allowed the bindable to change when further playlist items were added because mod select overlay shares the same global instance. To trigger this issue you'd add the classic mod first with one setting changed, and then reset the classic mod setting.

The above would fix the cause of the issue, however wouldn't fix an already-broken playlist item. In the broken state, the classic mod would arrive at the client containing one setting set to its default state (`true`). From the POV of both the client and server, this is still a correct state to be in even though the setting is redundant.
In order to fix this I've applied a second fix which converts the APIMods to local mods before comparing to the gameplay mods, rather than the inverse (converting gameplay mods to APIMods before comparing).

### CODE REVIEW

File: osu.Game/Online/API/APIMod.cs [hunk 1]
```diff
@@ -42,7 +42,7 @@ public APIMod(Mod mod)
                 var bindable = (IBindable)property.GetValue(mod);
 
                 if (!bindable.IsDefault)
-                    Settings.Add(property.Name.Underscore(), bindable);
+                    Settings.Add(property.Name.Underscore(), ModUtils.GetSettingUnderlyingValue(bindable));
```
Comment (frenzibyte) [inline osu.Game/Online/API/APIMod.cs:45]: One thing which spooks me about this is the "Difficulty Adjust" mod settings (whether it still works just fine after this change). Haven't got around testing that yet though.

**Comment by peppy** [comment]
One thing I think I noticed when looking into this (without your fix here) is that only the non-default settings seemed to be sent from (and to?) the server, is that the expectation?

File: osu.Game/Online/API/APIMod.cs (hunk 1 above)
Comment (smoogipoo) [reply osu.Game/Online/API/APIMod.cs:45]: Works fine as far as I can tell, why do you think it wouldn't?

**Comment by smoogipoo** [comment]
> only the non-default settings

Correct. I meant to bring that up because I'm not sure if it's correct. I feel like we should probably be sending all settings regardless, just to have forward-compatibility (if any setting defaults change?)

**Comment by peppy** [comment]
Yeah not sure. Have to consider the storage overhead of that (considerable).

File: osu.Game/Online/API/APIMod.cs (hunk 1 above)
Comment (frenzibyte) [reply osu.Game/Online/API/APIMod.cs:45]: Well with this change, the `Default` value will no longer be copied across to a new mod created by `ToMod`, but upon testing, it seems the refactor of the difficulty adjust mod settings are resilient enough to no longer rely on the `Default` value and `IsDefault` state of the setting. So it seems alright.

Comment (smoogipoo) [reply osu.Game/Online/API/APIMod.cs:45]: The check above already excludes default settings from being serialised and existed prior to this change.

Comment (frenzibyte) [reply osu.Game/Online/API/APIMod.cs:45]: I was referring to non-default settings with `Default` values populated to them, which used to be the case with difficulty adjust settings prior to the introduction of `DifficultyAdjustSettingControl`s.

Comment (smoogipoo) [reply osu.Game/Online/API/APIMod.cs:45]: Shouldn't be any difference. Bindable serialisation [goes via `Value`](https://github.com/ppy/osu-framework/blob/abe410f850c600c98b0787bba4c3a8de16a32c67/osu.Framework/Bindables/Bindable.cs#L404-L407) just as `ModUtils.GetSettingUnderlyingValue` does.

The method I used is even documented to be used for this purpose :smile:  :
https://github.com/ppy/osu/blob/7302e66c5f5706f6e774ece42a5da9d6fadc4667/osu.Game/Utils/ModUtils.cs#L157-L162

**Review by peppy — APPROVED** [review]

**Pull request merged**